client/docs/plugin_model.md

```diff
@@ -0,0 +1,169 @@
+are built in parts: a set of operations on a contract object, a
+set of commands that implement logical operations that span multiple
+invocations on multiple objects, and rarely a set of utility scripts.
+
+The contract operations portion of the plugin defines a set of
+objects, derived from the `contract_op_base` class, that each
+implement one of the methods on objects of a specific contract
+type. Often, these methods are shared with other contract types and
+can be "inherited" simply by importing them from another operations
+file. For example, the `kv-test` contract defines a `get` method. To
+invoke that method, the `kv-test` plugin defines a contract operation
+that sets up the invocation of the method on `kv-test` contract
+objects.
+
+The contract command portion of the plugin defines a set of commands
+that span multiple operations, dervied from the
+`contract_command_base` class. For example, provisioning a
+`token_issuer` contract object requires two operations to be performed
+(`add_endpoint` and `provision_token_issuer`). The `provision` command
+executes both of these in one command to simplify interaction with the
+`token_issuer` object.
+
+Scripts provide utility functions and are derived from the
+`script_command_base` class. Scripts are generally not associated with
+contract objects. The PDO client library provides scripts that
+simplify configuration of clients and communication with contract
+objects.
+
+All plugin classes define two important methods:
+* `add_arguments` -- adds commands for processing shell arguments with
+  `argparse`; the parameters can be processed through the pdo-shell or
+  other command line shells.
+* `invoke` -- invokes the operation on the contract object; command
+  line parameters that are required should be provided as positional
+  arguments and options arguments should be provided as keyword
+  arguments.
+
+All three types of plugin operations take the current interpreter
+`State` as a parameter to the `invoke` method.  The important
+difference between the different operations is that each type takes
+unique additional parameters. The `invoke` method in the
+`contract_op_base` takes an additional `Session` object with
+information about the enclave service connection that should be
+used. The `contract_command_base` uses a `Context` object to represent
+the configuration of a contract object and its relationship to other
+contract objects. Scripts, take no additional parameters beyond the
+`State`.
+
+Functionality defined in the pluging can be mapped to a `pdo-shell`
+command by defining an entry point and binding it to a command
+attribute in the shell class. Each type of operation (contract
+operations, contract commands and scripts) has a class-specific
+generator that creates a function for the shell,
+`create_shell_command`. The generated commands can be mapped into the
+`pdo-shell` using `bind_shell_command`.
+
+## Base Classes ##
```

Comment:
A short explanation as to the relationship/purpose of these base classes would be helpful. I read the previous paragraphs about the inputs to operations, commands and scripts, though the following descriptions are rather low-level describing implementation specifics rather than how I might use each class to build my contract.

Reply:
Feel free to provide some content :)